src/Mod/Path/PathTests/TestPathLevelMap.py

```diff
@@ -0,0 +1,170 @@
+# This is a standalone test, use it from command line.
+# This test will be rewritten as an automatic one and will be moved into the appropriate directory.
+# For visualization of test results the appropriate scilab scripts should be used.
+import PathLevelMap
+import math
+import numpy
+import random
+import time
+
+class TestPathLevelMap():
+    def sphere(self, x0, y0, z0, r, n ):
```

Comment:
Python unit tests always run in lexigraphical order within the test class.  
We usually name our tests something like test001() and then put a good description of the  test in the next comment line.  This way we have control over the order they execute.  This isn't strictly necessary but if the tests build on each other, it's good to bail out early rather than continuing.

Reply:
I don't understand how to include these tests in the main workflow. Some of them are significantly time-consuming. How much CPU time can I use for tests? Also, I have no idea how to run a test for the whole algorithm. Firstly, there are many possible combinations of parameters, and the second question is how to check the correctness of the created path. Can you give me some ideas?

Reply:
A lot of your tests seem to be stress tests.  I would not add those to the main test suite.  It's good to have them but we don't want to be that hard on the CI pipeline for every single PR.

Testing the whole algorithm is what I would call an 'integration test'.  That's also very difficult to write, especially  for user-facing functionality where, like you said, there are lots of combinations.

What we are striving for right now are unit tests.  These are tests that validate a single function and especially validate assumptions made by that function.  For example, if your function is expecting a list, will it return a ValueError if it gets a float or will it break later in the function with 'TypeError not iterable' messge?

Best Practices: https://www.testim.io/blog/unit-testing-best-practices/

I'm speaking in generalities here, not giving you advice on how to proceed.  Writing good tests is an acquired skill and one that I'm still acquiring.  So I don't feel like an expert.  I won't let that stop me.  :-)

If you don't have a pure function for the main algorithm, or such a function would take too many arguments, then consider factoring out some of the complexity to simpler functions and testing those.  

You can look at the generators I've already written and the corresponding unit tests.  I'm really happy with these.  Working this way changed how I write the code.  It forced me to make the code simpler, and more modular because it was easier to write the tests that way.  The side effect is that the code is easier to read and much more stable.  Also, I went faster because the effect was cumulative.  I'd get one feature and test working then start working on another feature.  if my changes broke the first feature, I knew it immediately.
  
Regarding the correctness of a generated path.  Start small.  Work up.
I would do a test case that uses a dirt simple model.  Maybe a sphere.  Does one waterline loop. Then compare the commands in the resulting path to your expected result.  You can get the string of a Path object with 
`obj.Path.toGCode()`